sips/cluster_consensus.md

```diff
@@ -0,0 +1,127 @@
+
+According to Monte Carlo simulations using a dataset based on the Mainnet, this proposal reduces to $21.60$% the current number of messages exchanged in the network. Note that this result includes aggregating the post-consensus messages into a single message.
+
+Regarding the number of bits exchanged, we estimate that this proposal will reduce the current value to, at least, $52.96$%. Notice that this reduction is not as significant as the number of messages reduction due to the larger post-consensus messages.
+
+Again with Monte Carlo simulations using the Mainnet dataset, the number of attestation duties aggregated presented the following distribution (notice that it also represents the number of partial signature messages merged into a single message).
+
+<p align="center">
+<img src="./images/cluster_consensus/aggregated_duties.png"  width="50%" height="10%">
+</p>
+
+## Spec changes
+
+### Design
+
+Currently, an operator manages many `Validator` objects. Each has a `DutyRunner` for a duty type, each with its own `QBFTController` with its unique ID (that is also inserted in its associated messages).
+
+<p align="center">
+<img src="./images/cluster_consensus/previous_scheme.drawio.png"  width="50%" height="10%">
+</p>
+
+For the proposed change to take place, different `Validator` objects should use the same `QBFT Instance`. For that, we propose decoupling the `QBFTController` object from the `DutyRunner`.
+
+<p align="center">
+<img src="./images/cluster_consensus/new_scheme.drawio.png"  width="50%" height="10%">
+</p>
+
+### New IDs
+
+Since a single `QBFT instance` will be responsible for several validator duties, its ID must not be dependent on a validator key but rather on a cluster of operators. For that, we propose changing the `MessageID` from
+
+```mermaid
+flowchart RL
+	subgraph MessageID
+		Domain
+		ValidatorPublicKey
+		Role
+	end
+```
+
+to
+
+
+```mermaid
+flowchart RL
+	subgraph MessageID
+		Domain
+		OperatorCluster
+		Role
+	end
+```
+
+### QBFT Controller
+
+The current `QBFTController` structure allows only one consensus instance at a time. This must change, extending the `QBFTController` into a router of messages for the different instances.
+
+### DutyRunner & QBFT Controller
+
+Since the `DutyRunner` will not have its specific `QBFTController`, it must have a way to start a consensus instance and receive its decided value. For that, we propose applying the observer design pattern by which the `DutyRunner` (observer) can be updated upon a `QBFT Instance` (observable) termination.
+
+We suggest that the `DutyRunner` holds a reference to the operator's `QBFTController` to start a `QBFT Instance` and observe it.
+
+```go
+func (r *DutyRunner) execute() {
+	r.QBFTController.StartConsensus(r.committee, r.duty, r)
+}
+
+func (c *QBFTController) StartConsensus(committee []types.Operator, duty types.Duty, observer Observer) {
+	instance := c.StartInstance(committee, duty)
+	instance.registerObserver(observer)
+}
+```
```

Comment:
> Since the `DutyRunner` will not have its specific `QBFTController`

This wording is confusing since you later show that the runner has a direct reference to `QBFTController`

Also we shouldn't get into implementation details with the observer unless absolutely necessary for tests

Reply:
I see why this is confusing. Sorry for that.

What I meant is that the `DutyRunner` used to have a `QBFTController` of its own. Now, the operator has the `QBFTController` and the `DutyRunner` will hold a reference to it.

I tried not to add a lot of code. I thought that would help to show how the observer pattern could be used.